walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,126 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Set;
+import java.util.HashSet;
+import java.util.ArrayList;
+import java.util.Comparator;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    Collection<String> mandatoryMembers = new ArrayList();
+    Collection<String> optionalMembers = new ArrayList();
+
+    // if meeting only has optional members treat them as mandatory
+    if (request.getAttendees().isEmpty()) {
+      mandatoryMembers = request.getOptionalAttendees();
+    } else {
+      mandatoryMembers = request.getAttendees();
+      optionalMembers = request.getOptionalAttendees(); 
+    }
+
+    // go through all events and find unavailable times
+    ArrayList<TimeRange> unavailableTimes = findUnavailableTimesMandatoryAttendees(events, mandatoryMembers);   
+
+    // find available times from unavilable times from mandatory members
+    ArrayList<TimeRange> availableTimes = findAvalibaleTimesMandatoryMembers(unavailableTimes, request);
```

Comment:
`findAvalibaleTimesMandatoryMembers` --> `findAvailableTImesMandatoryMembers` typo in available